File: forge/src/main/java/juuxel/adorn/platform/forge/block/SofaBlockForge.java
package juuxel.adorn.platform.forge.block;

import juuxel.adorn.block.SofaBlock;
import net.minecraft.world.level.block.state.BlockBehaviour.Properties;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.LevelReader;

public final class SofaBlockForge extends SofaBlock {
    public SofaBlockForge(Properties settings) {
        super(settings);
    }

    @Override
    public boolean isBed(BlockState state, BlockGetter level, BlockPos pos, LivingEntity sleeper) {
        return true;
    }

    @Override
    public Direction getBedDirection(BlockState state, LevelReader world, BlockPos pos) {
        var direction = getSleepingDirection(world, pos);
        return direction != null ? direction.getOpposite() : null;
    }

    @Override
    public void setBedOccupied(BlockState state, Level world, BlockPos pos, LivingEntity sleeper, boolean occupied) {
        super.setBedOccupied(state, world, pos, sleeper, occupied);
        var neighborPos = pos.relative(getSleepingDirection(world, pos, true));
        world.setBlockAndUpdate(neighborPos, world.getBlockState(neighborPos).setValue(OCCUPIED, occupied));
    }
}

File: forge/src/main/kotlin/juuxel/adorn/platform/forge/block/SofaBlockForge.kt
package juuxel.adorn.platform.forge.block

import juuxel.adorn.block.SofaBlock
import juuxel.adorn.block.variant.BlockVariant
import net.minecraft.block.BlockState
import net.minecraft.entity.Entity
import net.minecraft.entity.LivingEntity
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.world.BlockView
import net.minecraft.world.World
import net.minecraft.world.WorldView

class SofaBlockForge(variant: BlockVariant) : SofaBlock(variant) {
    override fun isBed(state: BlockState, world: BlockView, pos: BlockPos, player: Entity?): Boolean = true

    override fun getBedDirection(state: BlockState, world: WorldView, pos: BlockPos): Direction? =
        getSleepingDirection(world, pos)?.opposite

    override fun setBedOccupied(state: BlockState, world: World, pos: BlockPos, entity: LivingEntity, value: Boolean) {
        super.setBedOccupied(state, world, pos, entity, value)
        val neighborPos = pos.offset(getSleepingDirection(world, pos, ignoreNeighbors = true))
        world.setBlockState(neighborPos, world.getBlockState(neighborPos).with(OCCUPIED, value))
    }
}
